src/lib/merkle_mask/masking_merkle_tree.ml

```diff
@@ -57,10 +57,10 @@ module Make (Inputs : Inputs_intf.S) = struct
     { uuid = Uuid_unix.create ()
     ; parent = Error __LOC__
     ; detached_parent_signal = Async.Ivar.create ()
-    ; account_tbl = Location_binable.Table.create ()
-    ; token_owners = Token_id.Table.create ()
-    ; hash_tbl = Addr.Table.create ()
-    ; location_tbl = Account_id.Table.create ()
+    ; accounts = Location_binable.Map.empty
+    ; token_owners = Token_id.Map.empty
+    ; hashes = Addr.Map.empty
+    ; locations = Account_id.Map.empty
```

Comment:
@nholland94 one thing I noticed a while back is that comparing and hashing fields performs quite a bit of work (and allocations). That (among many other things) affects tables and maps that use `Account_id.t` keys. 

The hashing part is probably easily avoided by moving away from hash tables like in this PR (although any composite type that has field values inside it will still pay the cost when hashed), but if the `Field.compare` function is improved, that would still help here (for example, for a lookup that has to go through k=290 masks, the key being queried needs to be converted 290 amounts of times with the current implementation).

What do you think? could implementing `compare` as an `external` that directly compares the values in Rust help? I see that it is already used here when implementing the custom pointer type's `compare` operation handler

https://github.com/MinaProtocol/mina/blob/44af5c2865deb8b0363b8d976cb56f5624969b0c/src/lib/crypto/kimchi_bindings/stubs/src/arkworks/pasta_fp.rs#L31-L45

Reply:
Forgot the link to the functions I am talking about:

https://github.com/MinaProtocol/mina/blob/44af5c2865deb8b0363b8d976cb56f5624969b0c/src/lib/crypto/kimchi_backend/common/field.ml#L180-L197

Reply:
Nice catch. Yes, it should make a big difference to actually do the comparison in Rust instead of what we are currently doing in the kimchi backend. We have moved other data structures that were hashing many field elements to instead compare field elements, so this optimization would be very valuable overall.

Reply:
We want to be careful to preserve the behaviour: the above should convert to `into_repr` in rust before comparing, to ensure that we can still use it in the protocol and o1js.
More generally though, the function is badly implemented: it should use bitmasks, and also let rust return an array of `int`s that we can project to bigint instead of going bitwise, which will also basically improve bin_prot performance.

Reply:
The version I linked above is already using `into_repr()` internally in the implementation of `cmp`, but I may be missing some details related to o1js:

https://github.com/arkworks-rs/algebra/blob/v0.3.0/ff/src/fields/macros.rs#L513-L518

Btw, `caml_pasta_fp/fq_compare` already exist, so changing `compare` is a few lines diff (just created a very rough PR [here](https://github.com/MinaProtocol/mina/pull/14597)). I will get back to it tomorrow
 and look into improving `to_bignum_bigint`

Reply:
Regarding `to_bignum_bigint`, I see returning `int64` values is not currently supported by `ocaml-gen` but returning regular ints that with at-most 32 bits filled works. What is missing is a way to construct a bigint at once from all the input ints instead of having to create 8 temporary ones (from which then additional temporary bigints are created through shifts and `lor` operations). It is still better than going bit by bit, but still not ideal.

Something like this (incomplete and untested):

```ocaml
let combine_ints_to_bigint a b c d e f g h =
  let open Bignum_bigint in
  let a' = shift_left (of_int a) 224 in
  let b' = shift_left (of_int b) 192 in
  let c' = shift_left (of_int c) 160 in
  let d' = shift_left (of_int d) 128 in
  let e' = shift_left (of_int e) 96 in
  let f' = shift_left (of_int f) 64 in
  let g' = shift_left (of_int g) 32 in
  let h' = of_int h in
  a' lor b' lor c' lor d' lor e' lor f' lor g' lor h'
```

```rust
#[ocaml_gen::func]
#[ocaml::func]
pub fn caml_pasta_fp_to_ints(x: ocaml::Pointer<CamlFp>) -> Vec<i32> {
    let repr = x.as_ref().0.into_repr();
    let mut result: [i32; 8] = [0; 8];

    for (i, &value) in repr.0.iter().enumerate() {
        result[i * 2] = (value & 0xFFFFFFFF) as i32; // Lower 32 bits
        result[i * 2 + 1] = (value >> 32) as i32; // Upper 32 bits
    }

    result.to_vec()
}
````

Reply:
> What is missing is a way to construct a bigint at once from all the input ints instead of having to create 8 temporary ones (from which then additional temporary bigints are created through shifts and lor operations)

Purely with Zarith without intermediary bignums and shifts (`bignum.bigint` doesn't seem to expose `of_bits`), but according to the docs, the sign is not included in that conversion

```ocaml
let combine_ints_to_z a b c d e f g h =
  let buffer = Bytes.create 32 in
  Bytes.set_int32_le buffer 0 (Int32.of_int a);
  Bytes.set_int32_le buffer 4 (Int32.of_int b);
  Bytes.set_int32_le buffer 8 (Int32.of_int c);
  Bytes.set_int32_le buffer 12 (Int32.of_int d);
  Bytes.set_int32_le buffer 16 (Int32.of_int e);
  Bytes.set_int32_le buffer 20 (Int32.of_int f);
  Bytes.set_int32_le buffer 24 (Int32.of_int g);
  Bytes.set_int32_le buffer 28 (Int32.of_int h);
  Z.of_bits (Bytes.unsafe_to_string buffer)
```

Reply:
Just pushed a version (different to what was discussed here) to https://github.com/MinaProtocol/mina/pull/14599

It is a bit hackier in that it goes through bin_prot (because that is the only way exposed by `Bignum_bigint` to convert raw bytes into a bigint), but should be more efficient.